server/service/src/rnr_form/generate_rnr_form_lines.rs

```diff
@@ -1,95 +1,339 @@
+    let master_list_item_ids = get_master_list_item_ids(&ctx, master_list_id)?;
+
+    let period_length_in_days = get_period_length(&period);
+    let lookback_months = period_length_in_days as f64 / NUMBER_OF_DAYS_IN_A_MONTH;
+
+    // Get consumption/replenishment/adjustment stats for each item in the master list
+    let usage_by_item_map = get_usage_map(
+        &ctx.connection,
+        store_id,
+        Some(EqualFilter::equal_any(master_list_item_ids.clone())),
+        period_length_in_days,
+        &period.end_date,
     )?;
 
-    let item_ids = master_list_lines
+    // Get previous form data for initial balances
+    let previous_rnr_form_lines_by_item_id =
+        get_last_rnr_form_lines(&ctx.connection, previous_form.map(|f| f.rnr_form_row.id))?;
+
+    let rnr_form_lines = master_list_item_ids
         .into_iter()
-        .map(|line| line.item_id)
-        .collect();
-
-    let item_stats_rows = get_item_stats(
-        ctx,
-        &ctx.store_id,
-        None,
-        Some(ItemStatsFilter::new().item_id(EqualFilter::equal_any(item_ids))),
-    )?;
+        .map(|item_id| {
+            let initial_balance = get_opening_balance(
+                &ctx.connection,
+                previous_rnr_form_lines_by_item_id.get(&item_id),
+                store_id,
+                &item_id,
+                period.start_date,
+            )?;
 
-    let period_length = period
-        .end_date
-        .signed_duration_since(period.start_date)
-        .num_days() as f64;
+            let usage = usage_by_item_map.get(&item_id).copied().unwrap_or_default();
 
-    // TODO: oms store pref?? one day :)
-    let target_mos = 2;
+            let final_balance =
+                initial_balance + usage.replenished - usage.consumed + usage.adjusted;
 
-    let rnr_form_lines = item_stats_rows
-        .into_iter()
-        .map(|item_stat| {
-            // TODO consumption view needs prescriptions added!
-            // consumed over period
-            let quantity_consumed = 0.0;
+            let stock_out_duration = get_stock_out_duration(
+                &ctx.connection,
+                store_id,
+                &item_id,
+                period.end_date,
+                period_length_in_days as u32,
+                final_balance,
+            )?;
 
-            // evolution series (?) - points where soh is 0?
-            let stock_out_duration = 0;
+            let adjusted_quantity_consumed = get_adjusted_quantity_consumed(
+                period_length_in_days,
+                stock_out_duration as i64,
+                usage.consumed,
+            );
 
-            let time_in_stock = period_length - stock_out_duration as f64;
+            // This is only AMC for this period (if periods are monthly, this is redundant...)
+            // Should it be the default we have elsewhere... 3 months lookback?
+            let amc = usage.consumed / lookback_months;
```

Comment:
I.e. lookback months is just period_days/days_per_month...

Could use item_stats AMC, but that is current AMC for the item, rather than AMC at the time of R&R period...

Perhaps its fine for now, can expand on AMC in future?